### ISSUE TITLE
How to understand the relationship between tune.py and evaluate.py?

### ISSUE THREAD

**Issue by luozhengdong** [yandex-research/tabular-dl-tabr#16]
Hello,
My understanding is:
when tune.py finished, will get checkpoint.pt, DONE, report.json, summay.json, but only report.json and DONE(which is empty) are used by evaluate.py, which provide lr, weight_dalay,dropout parameters. The  checkpoint.pt is not used.

Usually train.py gets all those parameters, just trail=1 and the early stop mechanism is used, doesn't apply trail 100. 

So would you please explain the relationship between tune.py, evaluate.py,ensemble.py in more detail?

**Comment by Yura52** [comment]
Hello, the `bin/go.py` script is the best place to learn how `bin/tune.py`, `bin/evaluate.py` and `bin/ensemble.py` interact with each other.

For example, in a typical scenario, where the whole tune-evaluate-ensemble is performed, the following happens:
1. First, the hyperparameter tuning is performed: [link](https://github.com/yandex-research/tabular-dl-tabr/blob/75105013189c76bc4f247633c2fb856bc948e579/bin/go.py#L36)
2. Then, the evaluation: [link](https://github.com/yandex-research/tabular-dl-tabr/blob/75105013189c76bc4f247633c2fb856bc948e579/bin/go.py#L46)
3. Finally, the ensembling: [link](https://github.com/yandex-research/tabular-dl-tabr/blob/75105013189c76bc4f247633c2fb856bc948e579/bin/go.py#L47)

Does this help?

**Comment by Yura52** [comment]
> train, validate and test process

This process is simulated within `bin/ffn.py` for feed-forward networks, in `bin/tabr.py` for TabR, in `bin/xgboost_.py` for XGBoost, etc.

> Is tune.py to get better lr, weight_dalay and dropout parameters?

Yes, the only purpose of `bin/tune.py` is to tune hyperparameters.

> Is evaluate.py to get the results of an experiment with 15 random seeds?

Yes

> ensemble.py

Regarding `bin/ensemble.py`:
- First, it takes the 15 single model predictions produced by `bin/evaluate.py`
- Then, it splits them into 3 disjoint groups (that is, ensembles) consisting of 5 models each.
- One ensemble prediction == average prediction of single models in this ensemble.
- An example: [link](https://github.com/yandex-research/tabular-dl-tabr/tree/main/exp/tabr/california/0-ensemble-5)
